topic_message_query.go

```diff
@@ -64,79 +72,87 @@ func (query *TopicMessageQuery) GetEndTime() time.Time {
 }
 
 func (query *TopicMessageQuery) SetLimit(limit uint64) *TopicMessageQuery {
-	query.pb.Limit = limit
+	query.limit = &limit
 	return query
 }
 
 func (query *TopicMessageQuery) GetLimit() uint64 {
-	return query.pb.Limit
+	if query.limit != nil {
+		return *query.limit
+	} else {
+		return 0
+	}
 }
 
 func (query *TopicMessageQuery) SetErrorHandler(errorHandler func(stat status.Status)) *TopicMessageQuery {
 	query.errorHandler = errorHandler
 	return query
 }
 
+func (query *TopicMessageQuery) SetCompletionHandler(completionHandler func()) *TopicMessageQuery {
+	query.completionHandler = completionHandler
+	return query
+}
+
+func (query *TopicMessageQuery) SetRetryHandler(retryHandler func(err error) bool) *TopicMessageQuery {
+	query.retryHandler = retryHandler
+	return query
+}
+
 func (query *TopicMessageQuery) Subscribe(client *Client, onNext func(TopicMessage)) (SubscriptionHandle, error) {
-	ctx, cancel := context.WithCancel(context.TODO())
-	handle := newSubscriptionHandle(cancel)
+	handle := SubscriptionHandle{}
 
 	messages := sync.Map{}
 	messagesMutex := sync.Mutex{}
 
+	channel, err := client.mirrorNetwork.getNextMirrorNode().getChannel()
+	if err != nil {
+		return handle, err
+	}
+
 	go func() {
 		var subClient mirror.ConsensusService_SubscribeTopicClient
 		var err error
-		attempt := 0
-		resubscribe := true
-		channel, err := client.mirrorNetwork.getNextMirrorNode().getChannel()
-		if err != nil {
-			cancel()
-			grpcErr, ok := status.FromError(err)
-			if query.errorHandler == nil && ok {
-				query.errorHandler(*grpcErr)
-			}
-		}
 
 		for {
-			if resubscribe {
+			if query.attempt <= query.maxAttempts && subClient == nil {
```

Comment:
with the logic from line 138 - 145, it's possible that

`query.attempt > query.maxAttemps` and `subClient == nil`, the it falls to line 135, nil pointer deference

there the if condition should be just

`if subClient == nil {`

also have to change line 138